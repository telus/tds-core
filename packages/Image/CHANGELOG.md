# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 2.1.3 (2022-02-16)


### Bug Fixes

* gha release job npm publish token ([ce7aacd](https://github.com/telusdigital/tds/commit/ce7aacdd77e54c4f8e48eb31d0c05dc804530324))





## 2.1.2 (2022-02-15)


### Bug Fixes

* gha release job git config ([276ebb6](https://github.com/telusdigital/tds/commit/276ebb6968a0b56c9b87c178f6895a60ae108e71))





## 2.1.1 (2022-02-15)


### Bug Fixes

* gha release job git config ([1ab603d](https://github.com/telusdigital/tds/commit/1ab603d68c36219b0711fc353bc2515b64712ca9))





# [2.1.0](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.25...@tds/core-image@2.1.0) (2021-09-21)


### Features

* **core-image:** add react 17 to peer deps ([0386d63](https://github.com/telusdigital/tds/commit/0386d637dfbcc0f68aa0e17930db2f18eebf55ec))





## [2.0.25](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.24...@tds/core-image@2.0.25) (2021-01-15)

**Note:** Version bump only for package @tds/core-image





## [2.0.24](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.23...@tds/core-image@2.0.24) (2020-09-11)

**Note:** Version bump only for package @tds/core-image





## [2.0.23](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.22...@tds/core-image@2.0.23) (2020-09-08)


### Bug Fixes

* **core-image:** add support for styled-components v5 ([cc68788](https://github.com/telusdigital/tds/commit/cc68788f858dafe32f920f30b5ee463ca7a9ef7f))





## [2.0.22](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.21...@tds/core-image@2.0.22) (2020-09-04)

**Note:** Version bump only for package @tds/core-image





## [2.0.21](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.20...@tds/core-image@2.0.21) (2020-08-18)

**Note:** Version bump only for package @tds/core-image





## [2.0.20](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.19...@tds/core-image@2.0.20) (2020-02-03)

**Note:** Version bump only for package @tds/core-image





## [2.0.19](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.18...@tds/core-image@2.0.19) (2020-01-27)

**Note:** Version bump only for package @tds/core-image





## [2.0.18](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.17...@tds/core-image@2.0.18) (2020-01-23)

**Note:** Version bump only for package @tds/core-image





## [2.0.17](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.16...@tds/core-image@2.0.17) (2020-01-22)

**Note:** Version bump only for package @tds/core-image





## [2.0.16](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.15...@tds/core-image@2.0.16) (2019-10-09)

**Note:** Version bump only for package @tds/core-image





## [2.0.15](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.14...@tds/core-image@2.0.15) (2019-10-04)

**Note:** Version bump only for package @tds/core-image





## [2.0.14](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.13...@tds/core-image@2.0.14) (2019-09-12)

**Note:** Version bump only for package @tds/core-image





## [2.0.13](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.12...@tds/core-image@2.0.13) (2019-09-06)

**Note:** Version bump only for package @tds/core-image





## [2.0.12](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.11...@tds/core-image@2.0.12) (2019-08-22)

**Note:** Version bump only for package @tds/core-image





## [2.0.11](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.10...@tds/core-image@2.0.11) (2019-08-09)

**Note:** Version bump only for package @tds/core-image





## [2.0.10](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.9...@tds/core-image@2.0.10) (2019-08-01)

**Note:** Version bump only for package @tds/core-image





## [2.0.9](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.8...@tds/core-image@2.0.9) (2019-07-30)

**Note:** Version bump only for package @tds/core-image





## [2.0.8](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.7...@tds/core-image@2.0.8) (2019-06-25)

**Note:** Version bump only for package @tds/core-image





## [2.0.7](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.6...@tds/core-image@2.0.7) (2019-06-04)

**Note:** Version bump only for package @tds/core-image

## [2.0.6](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.5...@tds/core-image@2.0.6) (2019-05-07)

**Note:** Version bump only for package @tds/core-image

## [2.0.5](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.4...@tds/core-image@2.0.5) (2019-05-01)

**Note:** Version bump only for package @tds/core-image

## [2.0.4](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.3...@tds/core-image@2.0.4) (2019-05-01)

**Note:** Version bump only for package @tds/core-image

## [2.0.3](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.2...@tds/core-image@2.0.3) (2019-04-23)

**Note:** Version bump only for package @tds/core-image

## [2.0.2](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.1...@tds/core-image@2.0.2) (2019-04-22)

**Note:** Version bump only for package @tds/core-image

## [2.0.1](https://github.com/telusdigital/tds/compare/@tds/core-image@2.0.0...@tds/core-image@2.0.1) (2019-03-28)

**Note:** Version bump only for package @tds/core-image

# [2.0.0](https://github.com/telusdigital/tds/compare/@tds/core-image@1.0.1...@tds/core-image@2.0.0) (2019-03-19)

### Features

- **core-image:** upgrade to styled-components ([a251582](https://github.com/telusdigital/tds/commit/a251582))

### BREAKING CHANGES

- **core-image:** upgrade to styled-components

## [1.0.1](https://github.com/telusdigital/tds/compare/@tds/core-image@1.0.0...@tds/core-image@1.0.1) (2019-02-06)

**Note:** Version bump only for package @tds/core-image
