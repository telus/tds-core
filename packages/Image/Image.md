### Usage criteria

- We recommend not to `resize/crop` images in the browser.
- The width and height should be the same ratio as the natural width and height of the image.
- Image optimization should be done BEFORE rendering.
- You can use the `contentful API` to request an image of a certain size.
- `rounded="circle"` requires a square image `(width == height)` in order to generate a perfect circle.
- Non-square images will give a warning and will appear oval shaped.
- When using the image symbol in the Sketch file, replace the image in the overrides with and image with the same aspect ratio.

```jsx
<Image
  src="image-example.jpg"
  rounded="corners"
  width={300}
  height={300}
  alt="Image of co-workers collaborating"
/>
```
