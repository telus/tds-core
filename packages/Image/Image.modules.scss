.fluid {
  height: auto;
  max-width: 100%;
}
