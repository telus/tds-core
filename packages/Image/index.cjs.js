const Image = require('./dist/index.cjs')

module.exports = Image
