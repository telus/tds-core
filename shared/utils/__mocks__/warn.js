export const deprecate = jest.fn()
export const warn = jest.fn()
