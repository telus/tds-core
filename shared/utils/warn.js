/* eslint-disable import/prefer-default-export */

export const deprecate = (componentName, message) => {
  if (process.env.NODE_ENV === 'production') {
    return
  }

  console.warn(`[TDS] [Deprecate] ${componentName}: ${message}`) // eslint-disable-line no-console
}

export const warn = (componentName, message) => {
  if (process.env.NODE_ENV === 'production') {
    return
  }

  console.warn(`[TDS] ${componentName}: ${message}`) // eslint-disable-line no-console
}
