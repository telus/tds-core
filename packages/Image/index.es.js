import Image from './dist/index.es'

export default Image
