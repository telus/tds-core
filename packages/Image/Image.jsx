import React from 'react'
import PropTypes from 'prop-types'

import joinClasses from '../../shared/utils/joinClassNames'
import { warn } from '../../shared/utils/warn'
import safeRest from '../../shared/utils/safeRest'

import styles from './Image.modules.scss'
import borderStyles from '../../shared/styles/Borders.modules.scss'

/**
 * @version 1.0.0
 */
const Image = ({ src, width, height, alt, rounded, ...rest }) => {
  const isCircle = rounded === 'circle'
  const isSquare = width === height

  const classes = joinClasses(
    styles.fluid,
    rounded === 'corners' && borderStyles.rounded,
    isCircle && borderStyles.circular
  )

  if (isCircle && !isSquare) {
    warn(
      'Image',
      'rounded="circle" is not supported for non-square images. Please provide a square image, otherwise the resulting shape will not be a circle.'
    )
  }

  return (
    <img
      {...safeRest(rest)}
      src={src}
      width={width}
      height={height}
      alt={alt}
      className={classes}
    />
  )
}

Image.propTypes = {
  /**
   * The src attribute for the HTML img element.
   */
  src: PropTypes.string.isRequired,
  /**
   * The alt attribute for the HTML img element. Setting this attribute to an empty string (alt="") indicates that this image is not a key part of the content, and that non-visual browsers may omit it from rendering.
   */
  alt: PropTypes.string.isRequired,
  /**
   * The image's width.
   */
  width: PropTypes.number.isRequired,
  /**
   * The image's height.
   */
  height: PropTypes.number.isRequired,
  /**
   * Apply rounding.
   */
  rounded: PropTypes.oneOf(['circle', 'corners']),
}

Image.defaultProps = {
  rounded: undefined,
}

export default Image
