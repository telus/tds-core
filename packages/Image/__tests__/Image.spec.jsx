import React from 'react'
import { shallow } from 'enzyme'
import { warn } from '../../../shared/utils/warn'

import Image from '../Image'

jest.mock('../../../shared/utils/warn')

describe('Image', () => {
  const defaultProps = {
    src: 'test-image.jpg',
    alt: '',
    width: 200,
    height: 200,
  }

  const doShallow = (newProps = {}) => shallow(<Image {...defaultProps} {...newProps} />)

  it('renders', () => {
    const image = doShallow({})

    expect(image).toMatchSnapshot()
  })

  it('rounds 4 corners', () => {
    const image = doShallow({ rounded: 'corners' })

    expect(image).toHaveClassName('rounded')
  })

  describe('circular masking', () => {
    it('applies 50% border radius', () => {
      const image = doShallow({ rounded: 'circle' })

      expect(image).toHaveClassName('circular')
    })

    it('shows a warning when the image is not a square', () => {
      doShallow({ rounded: 'circle', width: 50, height: 60 })

      expect(warn).toHaveBeenCalled()
    })
  })

  it('passed additional attributes to image element', () => {
    const image = doShallow({ id: 'special-button', tabindex: 1 })
    expect(image).toHaveProp('id', 'special-button')
    expect(image).toHaveProp('tabindex', 1)
  })

  it('does not allow custom CSS', () => {
    const image = doShallow({ className: 'my-custom-class', style: { color: 'purple' } })
    expect(image).not.toHaveProp('className', 'my-custom-class')
    expect(image).not.toHaveStyle('color', 'purple')
  })
})
