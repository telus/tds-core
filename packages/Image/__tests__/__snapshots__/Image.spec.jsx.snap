// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders 1`] = `
<img
  alt=""
  className="fluid"
  height={200}
  src="test-image.jpg"
  width={200}
/>
`;
